### PR TITLE
Access Control: Add fine-grained access control to GET stats and settings handlers

### PR DESCRIPTION
**Pull request by kalleep**
**What this PR does / why we need it**:
Add fine-graind access control to stats and settings handler
Update admin navlinks to only include authorized links

### CODE REVIEW

File: pkg/api/index.go [hunk 1]
```diff
@@ -357,17 +357,23 @@ func (hs *HTTPServer) buildAdminNavLinks(c *models.ReqContext) []*dtos.NavLink {
 		adminNavLinks = append(adminNavLinks, &dtos.NavLink{
 			Text: "Orgs", Id: "global-orgs", Url: hs.Cfg.AppSubURL + "/admin/orgs", Icon: "building",
 		})
+		if hs.Cfg.PluginAdminEnabled {
+			adminNavLinks = append(adminNavLinks, &dtos.NavLink{
+				Text: "Plugin catalog", Id: "plugin-catalog", Url: hs.Cfg.AppSubURL + "/a/grafana-plugin-admin-app", Icon: "plug",
+			})
+		}
```
Comment (sakjur) [inline pkg/api/index.go:364]: I wonder if this will move the position of this in the menu? What would it take to have it below the two other checks?

**Review by sakjur — APPROVED** [review]
We've talked a little about having `settings:<section>:<key>` for these, and I think that would be a good idea. That said, I think this is a good start and we can move there fairly effortlessly.

File: pkg/api/admin.go [hunk 2]
```diff
@@ -19,3 +29,52 @@ func AdminGetStats(c *models.ReqContext) response.Response {
...
+func (hs *HTTPServer) getAuthorizedSettings(ctx context.Context, user *models.SignedInUser, bag setting.SettingsBag) (setting.SettingsBag, error) {
+	if hs.AccessControl.IsDisabled() {
+		return bag, nil
+	}
+
+	eval := func(scopes ...string) (bool, error) {
+		return hs.AccessControl.Evaluate(ctx, user, accesscontrol.ActionSettingsRead, scopes...)
+	}
+
+	ok, err := eval(accesscontrol.ScopeSettingsAll)
```
Comment (gamab) [inline pkg/api/admin.go:42]: Can you add an explanatory comment? Something like:
```suggestion
	// Early return if user has access to all sections
	ok, err := eval(accesscontrol.ScopeSettingsAll)
```

**Review by sakjur — APPROVED** [review]

File: pkg/api/admin.go [hunk 3]
```diff
@@ -19,3 +29,52 @@ func AdminGetStats(c *models.ReqContext) response.Response {
...
+		return nil, err
+	}
+	if ok {
+		return bag, nil
+	}
+
+	authorizedBag := make(setting.SettingsBag)
+
+	for section, keys := range bag {
+		ok, err := eval(getSettingsScope(section, "*"))
```
Comment (gamab) [inline pkg/api/admin.go:53]:
```suggestion
    	// Skip check on keys if user has access to all keys of the section
		ok, err := eval(getSettingsScope(section, "*"))
```

File: pkg/api/admin.go [hunk 4]
```diff
@@ -19,3 +29,52 @@ func AdminGetStats(c *models.ReqContext) response.Response {
 
 	return response.JSON(200, statsQuery.Result)
 }
+
+func (hs *HTTPServer) getAuthorizedSettings(ctx context.Context, user *models.SignedInUser, bag setting.SettingsBag) (setting.SettingsBag, error) {
```
Comment (gamab) [inline pkg/api/admin.go:33]: Nice and easy to read :+1:

**Review by gamab — APPROVED** [review]
Nice work!

**Pull request merged**